ext/cache/src/ext.rs

```diff
@@ -0,0 +1,96 @@
+// ext.rs
+
+// Copyright (C) 2020-2021 The Nitrocli Developers
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+use std::env;
+use std::ffi;
+use std::process;
+
+use anyhow::Context as _;
+
+#[derive(Debug)]
+pub struct Context {
+  pub nitrocli: ffi::OsString,
+  pub verbosity: Option<u8>,
+  pub project_dirs: directories::ProjectDirs,
+}
+
+impl Context {
+  pub fn from_env() -> anyhow::Result<Self> {
+    let nitrocli = env::var_os("NITROCLI_BINARY")
+      .context("NITROCLI_BINARY environment variable not present")
+      .context("Failed to retrieve nitrocli path")?;
+
+    let verbosity = env::var_os("NITROCLI_VERBOSITY")
+      .context("NITROCLI_VERBOSITY environment variablenot present")
+      .context("Failed to retrieve nitrocli verbosity")?;
+    let verbosity = if verbosity.len() == 0 {
+      None
+    } else {
+      let verbosity = verbosity
+        .to_str()
+        .context("Provided verbosity string is not valid UTF-8")?;
+      Some(u8::from_str_radix(verbosity, 10).context("Failed to parse verbosity")?)
+    };
+
+    let project_dirs = directories::ProjectDirs::from("", "", "nitrocli-cache")
+      .context("Could not determine the nitrocli-cache application directories")?;
+
+    Ok(Self {
+      nitrocli,
+      verbosity,
+      project_dirs,
+    })
+  }
+}
+
+#[derive(Debug)]
+pub struct Nitrocli {
+  cmd: process::Command,
+}
+
+impl Nitrocli {
+  pub fn from_context(ctx: &Context) -> Nitrocli {
+    let mut cmd = process::Command::new(&ctx.nitrocli);
+    if let Some(verbosity) = ctx.verbosity {
+      for _ in 0..verbosity {
+        cmd.arg("--verbose");
+      }
+    }
+    Self { cmd }
+  }
+
+  pub fn arg(&mut self, arg: impl AsRef<ffi::OsStr>) -> &mut Nitrocli {
+    self.cmd.arg(arg);
+    self
+  }
+
+  pub fn args<I, S>(&mut self, args: I) -> &mut Nitrocli
+  where
+    I: IntoIterator<Item = S>,
+    S: AsRef<ffi::OsStr>,
+  {
+    self.cmd.args(args);
+    self
+  }
+
+  pub fn text(&mut self) -> anyhow::Result<String> {
+    // TODO: inherit stderr from this process to show nitrocli debug messages
```

Comment:
I think it is.  Currently, stderr is just ignored for `text`.

Reply:
Hm, I was basing my comment on the fact that below we use `output.stderr` in the error. So that should do what we want it to, right? (whether we inherit stderr or read from it and print the output afterwards doesn't make much of a difference to me conceptually)

Reply:
Yes, but if the user invokes nitrocli with the `--verbose` option, they want to see the debug messages from the nitrocli instance invoked by the extension.   Currently they won’t be displayed because we ignore stderr (unless there is an error).

Reply:
Ah! True. Yeah, then we should address that TODO before checking in. Unfortunately then I think we can't capture the output to include it in errors programmatically. Otherwise I don't see how we can split the two concerns at this point.
```diff
--- ext/cache/src/ext.rs
+++ ext/cache/src/ext.rs
@@ -53,6 +53,12 @@ pub struct Nitrocli {
 impl Nitrocli {
   pub fn from_context(ctx: &Context) -> Nitrocli {
     let mut cmd = process::Command::new(&ctx.nitrocli);
+    // We want additional nitrocli emitted output to be visible to the
+    // user (typically controlled through -v/--verbose below). Note that
+    // this means that we will not be able to access this output for
+    // error reporting purposes.
+    cmd.stderr(process::Stdio::inherit());
+
     if let Some(verbosity) = ctx.verbosity {
       for _ in 0..verbosity {
         cmd.arg("--verbose");
@@ -76,15 +82,11 @@ impl Nitrocli {
   }

   pub fn text(&mut self) -> anyhow::Result<String> {
-    // TODO: inherit stderr from this process to show nitrocli debug messages
     let output = self.cmd.output().context("Failed to invoke nitrocli")?;
     if output.status.success() {
       String::from_utf8(output.stdout).map_err(From::from)
     } else {
-      Err(anyhow::anyhow!(
-        "nitrocli call failed: {}",
-        String::from_utf8_lossy(&output.stderr)
-      ))
+      Err(anyhow::anyhow!("nitrocli call failed"))
     }
   }

```
Not sure I like it, though.

Reply:
Alternatively, we could always capture stderr.  If the command succeeds, we print it to stderr, if it fails, we include it in the error message.  But as the error message will to end up in stderr, I don’t really see a benefit.

What exactly do you dislike?

Reply:
I think I dislike that we have two concerns that we are clubbing together and can't disentangle: the actual error (if there is such a thing) and tracing information. If we capture it then we have a potentially huge amount of (multi-line) data in an error that I am sure we eventually make the assumption is a one line thing.

Reply:
I see, but we only have the tracing information if the user explicitly asks for it.  I think that’s okay.

Reply:
Yeah. I'd suggest we stick to inheriting for now and don't expect programmatic access to errors (which I don't think we had plans of interpreting anyway, so they would always have been opaque, it's just a question of how we control surfacing them). Let's see when this bites us and then we will have to see what to do. What do you think? Do you insist on capturing?

Reply:
No, not at all!